### PR TITLE
Added tests to task manager 

### PR DESCRIPTION
**Pull request by schaable**
Part of https://github.com/NomicFoundation/hardhat/issues/5229

NOTES:
 - Read note at the top on why I'm testing this through `createHardhatRuntimeEnvironment` and not directly.
 - There is some duplication in the tests, but I avoided extracting it to keep things simple. 
 - Technically, we could repeat many of the tests with config tasks, but that seemed like overkill.

### CODE REVIEW

File: v-next/core/test/internal/tasks/task-manager.ts [hunk 1]
```diff
@@ -0,0 +1,908 @@
...
+ * TaskManagerImplementation.
+ */
+async function createHre({
+  config = {},
+  plugins = [],
+}: {
+  config?: HardhatUserConfig;
+  plugins?: HardhatPlugin[];
+  globalParameterMap?: GlobalParameterMap;
+} = {}) {
```
Comment (alcuadrado) [inline v-next/core/test/internal/tasks/task-manager.ts:37]: This function is not needed. Calling `createHardhatRuntimeEnvironment` has the same effect, or should have it. The third parameter shouldn't be used anywhere except for the CLI.

File: v-next/core/test/internal/tasks/task-manager.ts [hunk 2]
```diff
@@ -0,0 +1,908 @@
...
+    {},
+    {
+      resolvedPlugins,
+    },
+  );
+}
+
+describe("TaskManagerImplementation", () => {
+  it("should initialize the task manager with an empty set of tasks if no plugins or tasks are provided", async () => {
+    await assert.doesNotReject(createHre());
```
Comment (alcuadrado) [inline v-next/core/test/internal/tasks/task-manager.ts:51]: This test isn't really testing what its description does. At the same time, I don't think we have a way to actually test it, which makes me think that the `TaskManager` interface is incomplete. Maybe it should be able to either list all the tasks, or at least all the root tasks.

A similar problem will appear when printing the help messages.

Comment (alcuadrado) [reply v-next/core/test/internal/tasks/task-manager.ts:51]: Maybe add a `getTopLevelTasks()` in a separate PR?

File: v-next/core/test/internal/tasks/task-manager.ts [hunk 3]
```diff
@@ -0,0 +1,908 @@
...
+    const subsubtask1 = hre.tasks.getTask(["task1", "subtask1", "subsubtask1"]);
+    assert.deepEqual(subsubtask1.id, ["task1", "subtask1", "subsubtask1"]);
+    assert.equal(subsubtask1.pluginId, "plugin2");
+  });
+
+  /**
+   * These are all tested with plugin tasks, but the same logic applies to config tasks
+   */
+  describe("errors", () => {
+    it("should throw if there's a global parameter with the same name as a task named parameter", async () => {
```
Comment (alcuadrado) [inline v-next/core/test/internal/tasks/task-manager.ts:386]: Also positional

File: v-next/core/test/internal/tasks/task-manager.ts [hunk 4]
```diff
@@ -0,0 +1,908 @@
...
+            },
+          ],
+        }),
+        new HardhatError(HardhatError.ERRORS.TASK_DEFINITIONS.TASK_NOT_FOUND, {
+          task: "task1",
+        }),
+      );
+    });
+
+    it("should throw if trying to override a task and there is a name clash with an exising named parameter", async () => {
```
Comment (alcuadrado) [inline v-next/core/test/internal/tasks/task-manager.ts:556]: also with positional params

File: v-next/core/test/internal/tasks/task-manager.ts [hunk 5]
```diff
@@ -0,0 +1,908 @@
...
+          {
+            actorFragment: "Plugin plugin2 is",
+            namedParamName: "flag1",
+            task: "task1",
+          },
+        ),
+      );
+    });
+
+    it("should throw if a plugins tries to override a task defined in the config", async () => {
```
Comment (alcuadrado) [inline v-next/core/test/internal/tasks/task-manager.ts:836]: Nice test case

**Review by alcuadrado — APPROVED** [review]
LGTM. I left 3 comments, one of which should be addressed on a separate PR.

File: v-next/core/test/internal/tasks/task-manager.ts (hunk 3 above)
Comment (schaable) [reply v-next/core/test/internal/tasks/task-manager.ts:386]: Currently, it's only checking against named params, see https://github.com/NomicFoundation/hardhat/blob/6413a98470ba3c8671b41f5d90342ada56ea6cce/v-next/core/src/internal/tasks/task-manager.ts#L188

Does it also need to validate against positional parameters? If so, I can add the validation & tests in a follow-up PR.

Comment (alcuadrado) [reply v-next/core/test/internal/tasks/task-manager.ts:386]: Yes, i think it's better to be conservative. A follow up pr looks better.

**Pull request merged**